[SPARK-24522][UI] Create filter to apply HTTP security checks consistently.

vanzin opened the pull request:
Currently there is code scattered in a bunch of places to do different
things related to HTTP security, such as access control, setting
security-related headers, and filtering out bad content. This makes it
really easy to miss these things when writing new UI code.

This change creates a new filter that does all of those things, and
makes sure that all servlet handlers that are attached to the UI get
the new filter and any user-defined filters consistently. The extent
of the actual features should be the same as before.

The new filter is added at the end of the filter chain, because authentication
is done by custom filters and thus needs to happen first. This means that
custom filters see unfiltered HTTP requests - which is actually the current
behavior anyway.

As a side-effect of some of the code refactoring, handlers added after
the initial set also get wrapped with a GzipHandler, which didn't happen
before.

Tested with added unit tests and in a history server with SPNEGO auth
configured.

Commit:
[SPARK-24522][UI] Create filter to apply HTTP security checks consistently.

Currently there is code scattered in a bunch of places to do different
things related to HTTP security, such as access control, setting
security-related headers, and filtering out bad content. This makes it
really easy to miss these things when writing new UI code.

This change creates a new filter that does all of those things, and
makes sure that all servlet handlers that are attached to the UI get
the new filter and any user-defined filters consistently. The extent
of the actual features should be the same as before.

The new filter is added at the end of the filter chain, because authentication
is done by custom filters and thus needs to happen first. This means that
custom filters see unfiltered HTTP requests - which is actually the current
behavior anyway.

As a side-effect of some of the code refactoring, handlers added after
the initial set also get wrapped with a GzipHandler, which didn't happen
before.

Tested with added unit tests and in a history server with SPNEGO auth
configured.

SparkQA commented:
**[Test build #100051 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/100051/testReport)** for PR 23302 at commit [`9078ed7`](https://github.com/apache/spark/commit/9078ed79c63296507629e6feceedb0b699108d5b).

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/testing-k8s-prb-make-spark-distribution-unified/6040/
Test PASSed.

felixcheung commented:
Jenkins, retest this please

felixcheung commented:
was ` checking CRAN incoming feasibility ...Error in .check_package_CRAN_incoming(pkgdir) : 
  dims [product 24] do not match the length of object [0]`

SparkQA commented:
**[Test build #100132 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/100132/testReport)** for PR 23302 at commit [`9078ed7`](https://github.com/apache/spark/commit/9078ed79c63296507629e6feceedb0b699108d5b).

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/testing-k8s-prb-make-spark-distribution-unified/6116/
Test PASSed.

SparkQA commented:
**[Test build #100132 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/100132/testReport)** for PR 23302 at commit [`9078ed7`](https://github.com/apache/spark/commit/9078ed79c63296507629e6feceedb0b699108d5b).
 * This patch **fails due to an unknown error code, -9**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/100132/
Test FAILed.

gaborgsomogyi commented:
retest this, please

SparkQA commented:
**[Test build #100161 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/100161/testReport)** for PR 23302 at commit [`9078ed7`](https://github.com/apache/spark/commit/9078ed79c63296507629e6feceedb0b699108d5b).

SparkQA commented:
**[Test build #100161 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/100161/testReport)** for PR 23302 at commit [`9078ed7`](https://github.com/apache/spark/commit/9078ed79c63296507629e6feceedb0b699108d5b).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/100161/
Test FAILed.

SparkQA commented:
**[Test build #100182 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/100182/testReport)** for PR 23302 at commit [`9078ed7`](https://github.com/apache/spark/commit/9078ed79c63296507629e6feceedb0b699108d5b).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/100182/
Test FAILed.

gaborgsomogyi commented:
And another different flaky test.

gaborgsomogyi commented:
retest this, please

SparkQA commented:
**[Test build #100192 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/100192/testReport)** for PR 23302 at commit [`9078ed7`](https://github.com/apache/spark/commit/9078ed79c63296507629e6feceedb0b699108d5b).

SparkQA commented:
**[Test build #100192 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/100192/testReport)** for PR 23302 at commit [`9078ed7`](https://github.com/apache/spark/commit/9078ed79c63296507629e6feceedb0b699108d5b).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/100192/
Test PASSed.

Commit:
Merge branch 'master' into SPARK-24522

Commit:
Fix for adding the YARN AM filter in front of others + test.

SparkQA commented:
**[Test build #100290 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/100290/testReport)** for PR 23302 at commit [`514f3b1`](https://github.com/apache/spark/commit/514f3b1e5ceefc56c9634eaec2a3c77795b24a89).

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/testing-k8s-prb-make-spark-distribution-unified/6258/
Test PASSed.

SparkQA commented:
**[Test build #100290 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/100290/testReport)** for PR 23302 at commit [`514f3b1`](https://github.com/apache/spark/commit/514f3b1e5ceefc56c9634eaec2a3c77795b24a89).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/100290/
Test PASSed.

Commit:
Add some thread-safety when adding the YARN IP filter.

This might fix SPARK-26414 but I need to think a bit more about
that particular bug...

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/testing-k8s-prb-make-spark-distribution-unified/6283/
Test PASSed.

SparkQA commented:
**[Test build #100316 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/100316/testReport)** for PR 23302 at commit [`c5cded4`](https://github.com/apache/spark/commit/c5cded4f0fb5c616c2f5d116d22fd1281aa140f3).

Commit:
More tests on the YARN side.

SparkQA commented:
**[Test build #100318 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/100318/testReport)** for PR 23302 at commit [`76761cf`](https://github.com/apache/spark/commit/76761cfa4744b270b56d50667698f46f76c7a22b).

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/testing-k8s-prb-make-spark-distribution-unified/6285/
Test PASSed.

SparkQA commented:
**[Test build #100316 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/100316/testReport)** for PR 23302 at commit [`c5cded4`](https://github.com/apache/spark/commit/c5cded4f0fb5c616c2f5d116d22fd1281aa140f3).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/100316/
Test PASSed.

SparkQA commented:
**[Test build #100318 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/100318/testReport)** for PR 23302 at commit [`76761cf`](https://github.com/apache/spark/commit/76761cfa4744b270b56d50667698f46f76c7a22b).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/100318/
Test PASSed.

Commit:
Merge branch 'master' into SPARK-24522

vanzin commented:
@srowen @squito

SparkQA commented:
**[Test build #100654 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/100654/testReport)** for PR 23302 at commit [`34d6386`](https://github.com/apache/spark/commit/34d63869b47c01585fa50f4c1999d07e39422c3d).

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/testing-k8s-prb-make-spark-distribution-unified/6568/
Test PASSed.

srowen reviewed (approved):
Skimming it it all looked fine. It's an important thing to clean up and standardize.

SparkQA commented:
**[Test build #100654 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/100654/testReport)** for PR 23302 at commit [`34d6386`](https://github.com/apache/spark/commit/34d63869b47c01585fa50f4c1999d07e39422c3d).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/100654/
Test PASSed.

gaborgsomogyi commented on core/src/main/scala/org/apache/spark/ui/HttpSecurityFilter.scala:
```diff
@@ -0,0 +1,116 @@
+        }
+      }
+    } catch {
+      case e: IllegalArgumentException =>
```
If other type of exception is coming, will there be a response sent to the client?

vanzin replied:
You'll get an error 500.

gaborgsomogyi commented on core/src/main/scala/org/apache/spark/ui/JettyUtils.scala:
```diff
@@ -68,43 +69,16 @@ private[spark] object JettyUtils extends Logging {
+          val result = servletParams.responder(request)
+          response.getWriter.print(servletParams.extractFn(result))
         } catch {
           case e: IllegalArgumentException =>
```
Same question here.

gaborgsomogyi reviewed (commented):
Looks good, minor questions.

squito commented on core/src/main/scala/org/apache/spark/ui/HttpSecurityFilter.scala:
```diff
@@ -0,0 +1,116 @@
+    chain.doFilter(req, res)
+  }
+
+  private def isSafe(str: String): Boolean = stripXSS(str) == str
```
we didn't used to have this check.  Are you sure this is safe everywhere?  Eg. before if a param was `"runtime>10ms"` it would become `"runtime&gt;10ms"`, and with this change, it'll fail

vanzin replied:
I'll take a look, but I don't remember any pages that use that sort of syntax.

(I did find a separate issue with this filtering that I'll fix.)

vanzin replied:
It seems that while no pages I found use that syntax, the search boxes in the fancy task page (and probably other places) don't like this if you type an unexpected character.

I'll have to think about how to fix that and still keep this logic centralized...

squito replied:
I see, the reason you don't just update the params is that the `HttpServletRequest` doesn't let you?

The only workaround I can think of is to stuff the modified values into an attribute of the ServletContext, and then have a style rule to ban directly accessing them.

vanzin replied:
I can use an `HttpServletRequestWrapper` to do that, I just thought it would be better (and easier) to just error out instead of modifying the request. I'll go that route instead, at least we know the behavior will be the same and won't break something inadvertently...

squito commented on core/src/test/scala/org/apache/spark/ui/HttpSecurityFilterSuite.scala:
```diff
@@ -0,0 +1,159 @@
+      "alert3" -> """stdout'%2Balert(60)%2B'""",
+      "html" -> """stdout'"><iframe+id%3D1131+src%3Dhttp%3A%2F%2Fdemo.test.net%2Fphishing.html>"""
+    )
+    val badKeys = badValues.map { case (k, v) => v -> k }.toMap
```
could do `val badKeys = badValues.map(_.swap)`

Commit:
Remove header filtering.

This causes problems (e.g. a "Referer" that has a query string).

Commit:
Sanitize content instead of failing requests.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/testing-k8s-prb-make-spark-distribution-unified/6649/
Test PASSed.

SparkQA commented:
**[Test build #100756 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/100756/testReport)** for PR 23302 at commit [`66b776c`](https://github.com/apache/spark/commit/66b776ce55dfc584327a9efe946a34736d7cc0bb).

squito commented on core/src/test/scala/org/apache/spark/ui/HttpSecurityFilterSuite.scala:
```diff
@@ -0,0 +1,157 @@
+      assert(req.getParameter(k) != null)
+      assert(req.getParameter(k) != v)
+      assert(req.getParameterValues(k) != null)
+      assert(req.getParameterValues(k) != Array(v))
```
won't these pass just because its a different Array anyway?  eg `assert(Array(x) != Array(x))` always passes.  I think you could do `assert(req.getParameterValues(k).toSeq != Seq(v))`

vanzin replied:
Guess I was relying on Scala's `!=` being smarter.

srowen replied:
How about `===` and `!==` from scalatest? I think they do what you intend but haven't checked.

vanzin replied:
yeah, that works. Never used `!==` before.

srowen replied:
Really it's best to use these in tests, always; it generates better error messages like "foo wasn't true or "someComplexExpression was not less than 3"

squito reviewed (approved):
lgtm other than one small nit

Commit:
Compare seqs.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/testing-k8s-prb-make-spark-distribution-unified/6651/
Test PASSed.

Commit:
Use scalatest comparators in asserts.

SparkQA commented:
**[Test build #100759 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/100759/testReport)** for PR 23302 at commit [`68f3a52`](https://github.com/apache/spark/commit/68f3a52b7e5a9a1a1fca740bb4265c31c53b8adc).

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/testing-k8s-prb-make-spark-distribution-unified/6653/
Test PASSed.

SparkQA commented:
**[Test build #100756 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/100756/testReport)** for PR 23302 at commit [`66b776c`](https://github.com/apache/spark/commit/66b776ce55dfc584327a9efe946a34736d7cc0bb).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/100756/
Test FAILed.

SparkQA commented:
**[Test build #100758 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/100758/testReport)** for PR 23302 at commit [`3eb5956`](https://github.com/apache/spark/commit/3eb595693fe620ad4aef81c3dc87f3e3abbed066).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/100758/
Test FAILed.

SparkQA commented:
**[Test build #100759 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/100759/testReport)** for PR 23302 at commit [`68f3a52`](https://github.com/apache/spark/commit/68f3a52b7e5a9a1a1fca740bb4265c31c53b8adc).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/100759/
Test FAILed.

squito commented:
real failure in "org.apache.spark.ui.UISuite.add and remove handlers with custom user filter" ?

Commit:
Fix UISuite.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/testing-k8s-prb-make-spark-distribution-unified/6791/
Test PASSed.

SparkQA commented:
**[Test build #100901 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/100901/testReport)** for PR 23302 at commit [`13191fc`](https://github.com/apache/spark/commit/13191fcd707a0c4a634557e098a36d03587fc32f).

SparkQA commented:
**[Test build #100901 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/100901/testReport)** for PR 23302 at commit [`13191fc`](https://github.com/apache/spark/commit/13191fcd707a0c4a634557e098a36d03587fc32f).
 * This patch **fails to build**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/100901/
Test FAILed.

squito commented:
now looks like this needs to be updated after the mockito upgrade https://github.com/apache/spark/commit/e15a319ccd1125584c09c38ca90b252324df6998

vanzin commented:
Ah, was wondering about whether that was in. Will rebase later.

Commit:
Merge branch 'master' into SPARK-24522

Commit:
Fix code post merge.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/testing-k8s-prb-make-spark-distribution-unified/6800/
Test PASSed.

SparkQA commented:
**[Test build #100911 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/100911/testReport)** for PR 23302 at commit [`d185e12`](https://github.com/apache/spark/commit/d185e126a8f06cb3b577fd7622de5a87126e7f81).

SparkQA commented:
**[Test build #100911 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/100911/testReport)** for PR 23302 at commit [`d185e12`](https://github.com/apache/spark/commit/d185e126a8f06cb3b577fd7622de5a87126e7f81).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/100911/
Test PASSed.
